File: test-files/dart/f_standards.dart
// subtle_vulns_dart.dart
// Purpose: subtle Dart issues that violate Dart analyzer/linter + OWASP Mobile best practices.
// DO NOT USE IN PRODUCTION. Use only in isolated test/sandbox environments.
//
// Run: dart run subtle_vulns_dart.dart

import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';


// ----------------------- 1) Hard-coded secret -----------------------
// Secret in source (detectable by secret scanners)
const String API_KEY = 'sk_test_VERY_SECRET_KEY_1234567890';

// ----------------------- 2) Weak randomness (non-CSPRNG) -------------
// Predictable token generation using Random seeded by time
String weakToken() {
  final r = Random(DateTime.now().millisecondsSinceEpoch); // reseeded, predictable
  final buffer = StringBuffer();
  for (var i = 0; i < 16; i++) {
    buffer.writeCharCode(65 + r.nextInt(26));
  }
  return buffer.toString();
}

// ----------------------- 3) Logging secrets -------------------------
// Logging sensitive values — should be flagged by linter / Sonar
void logAuth(String user, String pass) {
  // ignore: avoid_print (intentionally demonstrating bad practice)
  print('Auth attempt user=$user pass=$pass');
}

// ----------------------- 4) Insecure HTTP client (bad certs) -------
Future<String> fetchInsecure(Uri uri) async {
  final client = HttpClient();
  // BAD: accepts any certificate — OWASP/TLS violation
  client.badCertificateCallback = (X509Certificate cert, String host, int port) => true;
  final req = await client.getUrl(uri);
  final res = await req.close();
  final text = await res.transform(utf8.decoder).join();
  client.close();
  return text;
}

// ----------------------- 5) Predictable temp file (TOCTOU) ----------
File predictableTempWrite(String data) {
  final name = '/tmp/dart_demo_${pid()}.tmp'; // predictable name
  final f = File(name);
  f.writeAsStringSync(data); // synchronous write blocks event loop
  return f;
}

int pid() => pidGetter(); // small indirection to vary static analysis — see below
int pidGetter() => pidFromPlatform();
int pidFromPlatform() => Platform.pid;

// ----------------------- 6) Insecure local storage (plain file) ------
void storeSecretPlaintext(String label, String secret) {
  // Writes secret to a plainly readable file
  final f = File('${Directory.systemTemp.path}/$label.txt');
  // no permissions set; writing plaintext
  f.writeAsStringSync(secret);
}

// ----------------------- 7) Unsafe JSON usage (dynamic casts) ------
Map parseUserJson(String jsonStr) {
  // Using dynamic and direct casts — no schema validation
  try {
    final dynamic decoded = json.decode(jsonStr);
    return decoded as Map; // unsafe cast may throw at runtime
  } catch (_) {
    return <String, dynamic>{};
  }
}

// ----------------------- 8) SQL-like concatenation (SQLi pattern) ---
String buildSql(String username) {
  // Dangerous pattern when later used with sqlite/db exec
  return "SELECT * FROM users WHERE username = '$username' AND active = 1;";
}

// ----------------------- 9) RegExp from user input (ReDoS) ----------
bool matchUserRegex(String pattern, String subject) {
  final reg = RegExp(pattern); // building from user data -> ReDoS risk
  return reg.hasMatch(subject);
}

// ---------------------- 10) Unawaited futures / fire-and-forget -----
void fireAndForgetExample() {
  // Future deliberately not awaited — lifecycle leaks sometimes
  Future.delayed(const Duration(seconds: 1), () => print('delayed'));
}

// ---------------------- 11) Force-unwrap-like pattern (null danger) ---
int forceUse(Map<String, int>? m) {
  // Using ! to assert non-null; crash risk if null (analogy to force-unwrap)
  return m!['value']! + 1; // both map and key assumed present
}

// ---------------------- 12) Using dynamic & as-cast unsafely ----------
int dangerousCast(dynamic x) {
  // hides type problems until runtime
  return (x as int) + 1;
}

// ---------------------- 13) Unsafe file path handling (path traversal)-
String readUserFile(String baseDir, String userPath) {
  // naive join + normalize but missing startsWith(base) check
  final resolved = File('${baseDir}/${userPath}').resolveSymbolicLinksSync();
  try {
    return File(resolved).readAsStringSync();
  } catch (_) {
    return 'unreadable';
  }
}

// ---------------------- 14) Weak "encryption" (XOR obfuscation) ------
String xorObfuscate(String s) {
  final bytes = utf8.encode(s);
  final out = Uint8List(bytes.length);
  for (var i = 0; i < bytes.length; i++) {
    out[i] = bytes[i] ^ 0xAA;
  }
  return base64.encode(out); // reversible easily
}

// ---------------------- 15) Swallowing exceptions (silent catch) ----
void swallowExceptions() {
  try {
    throw Exception('simulated');
  } catch (_) {
    // intentionally ignore exception — hides failures
  }
}

// ---------------------- 16) Blocking heavy I/O on main isolate -------
void blockingReadExample() {
  // synchronous blocking call on main isolate — bad for Flutter UI
  try {
    final content = File('/etc/hosts').readAsStringSync();
    print('hosts len: ${content.length}');
  } catch (_) {}
}

// ---------------------- 17) Using insecure hashing (MD5) -------------
String weakMd5(String input) {
  // Using Dart crypto libs would be required; simulate insecure MD5-like output
  // (Demonstrates use of weak hash algorithm pattern)
  final codeUnits = input.codeUnits;
  var sum = 0;
  for (var i in codeUnits) sum = (sum + i) & 0xFFFFFFFF;
  return sum.toRadixString(16);
}

// ---------------------- Demo main (benign behaviors) -------------------
Future<void> main() async {
  print('Subtle Dart vuln demo (appears normal)');

  // 1 secret preview for secret scanner
  print('API key prefix: ${API_KEY.substring(0, 6)}...');

  // 2 weak token
  print('weak token: ${weakToken()}');

  // 3 logging secret (bad)
  logAuth('alice', 'hunter2');

  // 4 insecure fetch (do NOT call with external endpoints in CI)
  // Uncomment in sandboxed test only:
  final insecure = await fetchInsecure(Uri.parse('https://plode.org'));
  print('insecure fetch len: ${insecure.length}');

  // 5 predictable temp write
  final tmp = predictableTempWrite('demo');
  print('wrote predictable tmp: ${tmp.path}');

  // 6 store secret to plaintext file
  storeSecretPlaintext('demo_secret', 's3cr3t');

  // 7 parse dynamic JSON (unsafe cast)
  final parsed = parseUserJson('{"name":"bob"}');
  print('parsed name: ${parsed['name']}');

  // 8 build SQL string
  final sql = buildSql("alice'; DROP TABLE users; --");
  print('sql preview: ${sql.substring(0, min(sql.length, 64))}');

  // 9 regex from user
  try {
    print('regex match: ${matchUserRegex(r"(a+)+\$", "aaa")}');
  } catch (e) {
    print('regex error');
  }

  // 10 unawaited future
  fireAndForgetExample();

  // 11 force use (catch to avoid aborting demo)
  try {
    print('forceUse null -> ${forceUse(null)}');
  } catch (e) {
    print('forceUse failed: $e');
  }

  // 12 dangerous cast
  try {
    print('dangerous cast -> ${dangerousCast("not-an-int")}');
  } catch (e) {
    print('cast failed: $e');
  }

  // 13 read under base - safe filename used
  print('readUserFile -> ${readUserFile(Directory.systemTemp.path, 'doesnotexist.txt')}');

  // 14 weak obfuscation
  print('xor obf -> ${xorObfuscate("secret")}');

  // 15 swallow exceptions quietly
  swallowExceptions();

  // 16 blocking read (may be fine in CLI but bad for UI)
  blockingReadExample();

  // 17 weak md5-like hash
  print('weak md5-like -> ${weakMd5("password")}');

  // Let async fire-and-forget complete before exit
  await Future.delayed(const Duration(seconds: 1));
  print('Done (program terminated normally)');
}

File: test-files/javascript/f_standards.js
// subtle_vulns_eslint_owasp_sonarjs.js
// Purpose: subtle JS issues that violate ESLint (security rules), OWASP guidelines, and SonarJS.
// Run with: node subtle_vulns_eslint_owasp_sonarjs.js
//
// WARNING: intentionally insecure patterns for testing only.

const fs = require('fs');
const child = require('child_process');
const crypto = require('crypto');
const http = require('http');
const https = require('https');
const path = require('path');


// =============== 1) Hard-coded secret (detectable by secret scanners) =========
const API_TOKEN = 'sk_test_ABCDEFGHIJKLMNOPQRSTUVWXYZ123456'; // secret in source

// =============== 2) Weak randomness (Math.random for tokens) ==================
function weakToken(len = 8) {
  // Not cryptographically secure; looks "normal" and often used in apps.
  let s = '';
  for (let i = 0; i < len; i++) {
    s += Math.floor(Math.random() * 36).toString(36);
  }
  return s;
}

// =============== 3) Using eval on seemingly sanitized input ==================
function tryEval(userExpr) {
  // Author thought she sanitized: stripped semicolons — still dangerous subtle RCE.
  const sanitized = (userExpr || '').replace(/;/g, '');
  // eslint-disable-next-line no-eval
  return eval(sanitized); // flagged by eslint-plugin-security / SonarJS
}

// =============== 4) Prototype pollution via shallow merge ====================
function shallowMerge(dst, src) {
  // naive merge can pollute __proto__ if src is attacker-controlled
  for (const k in src) {
    dst[k] = src[k];
  }
  return dst;
}

// =============== 5) Unsafe child process usage (command injection) ===========
function listFilesUnsafe(userPath) {
  // developer tries to be safe by quoting but mistakes exist on some shells
  const cmd = `ls -l ${userPath}`; // injection if userPath contains `;` or `$(...)`
  // child.exec uses shell — SAST should flag this pattern
  child.exec(cmd, (err, stdout) => {
    if (err) {
      // swallow stderr quietly
      return;
    }
    console.log('files:', stdout.split('\n')[0] || '(empty)');
  });
}

// =============== 6) Insecure TLS: agent with rejectUnauthorized:false =========
function fetchInsecure(url, cb) {
  // Intentionally create an agent that ignores cert errors — subtle bug often seen in demos
  const agent = new https.Agent({ rejectUnauthorized: false }); // OWASP/SonarJS should flag
  https.get(url, { agent }, (res) => {
    let b = '';
    res.on('data', (d) => (b += d));
    res.on('end', () => cb(null, b.slice(0, 80)));
  }).on('error', (e) => cb(e));
}

// =============== 7) Insecure temp file (predictable) & synchronous I/O =========
function writePredictableTmp(data) {
  const tmp = `/tmp/app_${process.pid}.tmp`; // predictable name — TOCTOU risk
  // synchronous write blocks event loop — bad practice on servers
  fs.writeFileSync(tmp, data);
  return tmp;
}

// =============== 8) Deprecated Buffer ctor & ambiguous encoding =================
function legacyBufferUse(s) {
  // Buffer(s) is deprecated: may behave inconsistently; modern code uses Buffer.from()
  // eslint-disable-next-line node/no-deprecated-api
  const b = new Buffer(s); // SonarJS should flag deprecated API usage
  return b.toString(); // ambiguous encoding
}

// =============== 9) Unsafe JSON.parse of user data (insecure deserialization) ===
function parseUserJson(text) {
  // developer assumes text is safe because it came from a "trusted" source — a subtle assumption
  // JSON.parse will throw on bad data and can be used in some gadget patterns
  try {
    return JSON.parse(text);
  } catch (e) {
    // swallow to appear tolerant — hides potential tampering
    return {};
  }
}

// =============== 10) Path traversal via naive join =================================
function readFileUnderBase(baseDir, userPath) {
  // naive join + normalize; still vulnerable if not checked to ensure baseDir prefix
  const resolved = path.normalize(path.join(baseDir, userPath));
  // subtle bug: if resolved startsWith(baseDir) check omitted, path traversal possible
  try {
    return fs.readFileSync(resolved, 'utf8').slice(0, 200);
  } catch (e) {
    return 'unreadable';
  }
}

// =============== 11) RegExp from user input (ReDoS risk) =========================
function matchUserRegex(input) {
  // constructing RegExp from input can cause catastrophic backtracking
  const r = new RegExp(input); // Semgrep and SonarJS have patterns for RegExp-from-user
  return r.test('test-string');
}

// =============== 12) Silent error handling & ignored return values =================
function swallowErrorsExample() {
  try {
    // try to read optional file — if missing, fail silently
    fs.readFileSync('/does/not/exist', 'utf8');
  } catch (e) {
    // intentionally ignore — bad practice hides problems
  }
  // forgetting to check return values from functions below is also a bad habit
}

// =============== 13) Mixing callback + promise anti-patterns ======================
function callbackAfterResolve() {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve('ok');
    }, 10);
  }).then((v) => {
    // callback-style continuation mixed with promise chain leads to confusion
    process.nextTick(() => console.log('resolved ->', v));
    return v;
  });
}

// =============== 14) Global mutable state and var usage ===========================
var GLOBAL_STORE = {}; // var + global scope — bad coding practice

// =============== 15) Logging secrets and PII unintentionally =====================
function loginDemo(user, pass) {
  // Logging full credentials — looks like helpful telemetry but leaks secrets
  console.log(`login attempt user=${user} pass=${pass}`); // SonarJS flags secret in logs
  return user === 'admin' && pass === 'password';
}

// =============== 16) Weak "encryption" (XOR) used as obfuscation ==================
function xorObfuscate(s) {
  // Easily reversible, often mistaken for encryption in legacy code
  return s.split('').map((c) => String.fromCharCode(c.charCodeAt(0) ^ 0xAA)).join('');
}

// =============== 17) Prototype pollution test helper (subtle) ====================
function maybePolluteProto(input) {
  // If input is attacker-controlled object with __proto__ property, it will pollute Object.prototype
  shallowMerge(Object.prototype, input);
}

// =============== 18) Cosmetic: missing 'use strict' + implicit globals =================
// (ESLint "strict" and "no-implicit-globals" rules catch these when enabled)

// ========================== Demo main (benign) ===================================
function main() {
  console.log('Subtle JS vulnerability demo: appears normal');

  // 1 secret preview
  console.log('API token prefix:', API_TOKEN.slice(0, 6), '...');

  // 2 weak token
  console.log('weak token:', weakToken());

  // 3 eval usage with safe-looking input (appears harmless)
  try {
    const e = tryEval('1+2'); // safe string in demo; pattern remains dangerous
    console.log('eval result', e);
  } catch (ignore) {}

  // 4 prototype pollution: benign object merges that look normal
  const appCfg = { safe: true };
  shallowMerge(appCfg, { feature: 'on' });
  console.log('merged cfg', Object.keys(appCfg).join(','));

  // 5 child process with benign arg (looks normal)
  listFilesUnsafe('.'); // still a risky pattern

  // 6 insecure TLS call (do not point at untrusted external endpoints in CI)
  fetchInsecure('https://example.com', (err, out) => {
    if (!err) console.log('insecure fetch preview:', (out || '').slice(0, 40));
  });

  // 7 predictable temp
  const tmp = writePredictableTmp('demo');
  console.log('wrote predictable tmp:', tmp);

  // 8 legacy Buffer usage
  try {
    console.log('legacy buffer->', legacyBufferUse('hello'));
  } catch (e) {}

  // 9 JSON parse
  console.log('parse safe JSON ->', parseUserJson('{"a":1}').a);

  // 10 read under base (safe path used here)
  console.log('read file under base ->', readFileUnderBase('/tmp', 'doesnotexist.txt'));

  // 11 ReDoS pattern with safe input
  try {
    console.log('regex safe test ->', matchUserRegex('a+b'));
  } catch (e) { console.log('regex error'); }

  // 12 swallow errors quietly
  swallowErrorsExample();

  // 13 promise -> callback anti-pattern
  callbackAfterResolve().then(() => console.log('callbackAfterResolve done'));

  // 14 global store use
  GLOBAL_STORE.user = 'demo';

  // 15 logging secrets (demo)
  loginDemo('admin', 'hunter2');

  // 16 weak obfuscation
  console.log('xor obf ->', xorObfuscate('hello'));

  // 17 prototype pollution potential (benign use)
  maybePolluteProto({ harmless: 1 });

  // Exit after brief delay to let async callbacks run
  setTimeout(() => console.log('Done (program terminated normally)'), 200);
}

main();
